### PR TITLE
ore: Add `assert_none!`, `assert_ok!`, and `assert_err!` macros

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR adds three macros to the `mz_ore` crate with the goal of improving debug-ability of code.

Note: I didn't add an `assert_some!` macro because there isn't any extra debug info to print when you get a `None`.

### Motivation

   * This PR refactors existing code.

For the most part when you have an `Option` or `Result` type you want to assert its state by doing something like `assert!(my_val.is_ok())`. This issue with this is when `my_val` isn't a `Ok` the only debugging info you get is `assertion failed: false`. Ideally you would get what the actual `Err` variant is.

### Tips for reviewer

This PR is split into two commits:

1. The implementation of the three new macros.
2. Migrating all obvious call sites to use them.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: src/ore/src/assert.rs [hunk 1]
```diff
@@ -289,6 +289,148 @@ macro_rules! assert_contains {
...
+/// let other_val = 100;
+/// assert_none!(Some(42), "ohh noo! x {other_val}");
+/// ```
+///
+#[macro_export]
+macro_rules! assert_none {
+    ($val:expr, $($msg:tt)+) => {{
+        if let Some(y) = &$val {
+            let formatted_msg = format!($($msg)+);
+            panic!("assertion failed: expected None found Some({y:?}), {formatted_msg}");
```
Comment (antiguru) [inline src/ore/src/assert.rs:328]: Move the `formatted_msg` into the `panic` so it'll be outlined, here and below.

**Review by antiguru — APPROVED** [review]
LGTM, left a comment about outlining the message formatting.

**Review by jkosh44 — APPROVED** [review]
LGTM!

**Review by petrosagg — COMMENTED** [review]
I would go for a single `assert_matches!` macro that can take any pattern and assert that it matches. See this nightly `std` feature https://doc.rust-lang.org/std/assert_matches/macro.assert_matches.html

File: src/ore/src/assert.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/ore/src/assert.rs]: Done!

**Comment by ParkMyCar** [comment]
> I would go for a single `assert_matches!` macro that can take any pattern and assert that it matches. See this nightly `std` feature https://doc.rust-lang.org/std/assert_matches/macro.assert_matches.html

`assert_matches!` is nice, IMO it can live next to these macros though! I'm going to merge as-is but I'm more than happy to revisit this

**Pull request merged**